libsolidity/formal/Predicate.cpp

```diff
@@ -214,19 +219,79 @@ string Predicate::formatSummaryCall(
 	}
 
 	/// The signature of a function summary predicate is: summary(error, this, abiFunctions, cryptoFunctions, txData, preBlockChainState, preStateVars, preInputVars, postBlockchainState, postStateVars, postInputVars, outputVars).
-	/// Here we are interested in preInputVars to format the function call,
-	/// and in txData to retrieve `msg.value`.
+	/// Here we are interested in preInputVars to format the function call.
+
+	Json::Value txModel;
 
-	string value;
-	if (auto v = readTxVars(_args.at(4)).at("msg.value"))
+	if (_appendTxVars)
 	{
-		bigint x(*v);
-		if (x > 0)
-			value = "{ value: " + *v + " }";
+		set<string> txVars;
+		if (isFunctionSummary())
+		{
+			solAssert(programFunction(), "");
+			if (programFunction()->isPayable())
+				txVars.insert("msg.value");
+		}
+		else if (isConstructorSummary())
+		{
+			FunctionDefinition const* fun = programFunction();
+			if (fun && fun->isPayable())
+				txVars.insert("msg.value");
+		}
+
+		struct TxVarsVisitor: public ASTConstVisitor
+		{
+			bool visit(MemberAccess const& _memberAccess)
+			{
+				Expression const* memberExpr = SMTEncoder::innermostTuple(_memberAccess.expression());
+
+				Type const* exprType = memberExpr->annotation().type;
+				solAssert(exprType, "");
+				if (exprType->category() == Type::Category::Magic)
+					if (auto const* identifier = dynamic_cast<Identifier const*>(memberExpr))
+					{
+						ASTString const& name = identifier->name();
+						if (name == "block" || name == "msg" || name == "tx")
+							txVars.insert(name + "." + _memberAccess.memberName());
+					}
+
+				return true;
+			}
+
+			set<string> txVars;
+		} txVarsVisitor;
+
+		if (auto fun = programFunction())
+		{
+			fun->accept(txVarsVisitor);
+			txVars += txVarsVisitor.txVars;
+		}
+
+		// Here we are interested in txData from the summary predicate.
+		auto txValues = readTxVars(_args.at(4));
+		vector<string> values;
+		for (auto const& _var: txVars)
+			if (auto v = txValues.at(_var))
+				txModel[_var] = *v;
+	}
+
+	string txModelStr;
+	if (!txModel.empty())
+	{
+		// We can either have a pretty version which has line breaks,
+		// or a compact version that does not but also has no spaces
+		// and has quotes.
+		// So we take the pretty version and remove line breaks,
+		// replace double spaces (which are left after removing line breaks)
+		// by single spaces, and remove quotes.
+		txModelStr = util::jsonPrettyPrint(txModel);
+		boost::replace_all(txModelStr, "\n", "");
+		boost::replace_all(txModelStr, "  ", " ");
+		boost::replace_all(txModelStr, "\"", "");
```

Comment:
@hrkrshnn this gives me what I want. Let me know if you prefer this or the previous version

Reply:
Actually it doesn't. The end of the string is then `name: value}`, whereas it should be `name: value }`. It's getting too hacky, I'm reverting